java/arcs/core/host/AbstractArcHost.kt

```diff
@@ -10,43 +10,239 @@
+            particleContext.handles[handleSpec.key] = handle
+        }
+
+        return particleContext
+    }
+
+    fun lookupParticleContextOrCreate(
+        context: ArcHostContext,
+        spec: Plan.Particle,
+        particle: Particle
+    ) = context.particles[spec]?.copy(particle = particle) ?: ParticleContext(particle)
+
+    suspend fun performLifecycleForContext(context: ArcHostContext) {
+        for (particleContext in context.particles.values) {
+            performParticleLifecycle(particleContext)
+        }
+    }
+
+    suspend fun performParticleLifecycle(particleContext: ParticleContext) {
+        if (particleContext.particleState == ParticleState.Instantiated) {
+            particleContext.particle.onCreate()
+            particleContext.particleState = ParticleState.Created
+        }
+
+        // TODO: verify this won't be true if the host crashes
+        if (particleContext.particleState == ParticleState.Started) {
+            // particle already started
+            return
+        }
+
+        // If we reach here, particle is being restarted
+        if (particleContext.particleState == ParticleState.Stopped) {
+            particleContext.particleState = ParticleState.Created
+        }
+
+        // This is temporary until the BaseParticle PR lands and onStartup() API lands.
+        // We force sync() calls in lieu of onStartup() API for demos
+        if (particleContext.particleState == ParticleState.Created) {
+            for ((handleName, handle) in particleContext.handles) {
+                particleContext.run {
+                    particle.onHandleSync(handle, false)
+                }
+            }
+            particleContext.particleState = ParticleState.Started
+        }
+    }
+
+    suspend fun maybeRequestResurrection(context: ArcHostContext) {
+        if (context.arcState == ArcState.NeverStarted) {
+            // TODO: First time, request resurrection for context.
+        }
+    }
+
+    suspend fun maybeCancelResurrection(context: ArcHostContext) {
+        // TODO: wire up resurrection cancellation
+    }
+
+    /**
+     * Given a handle name, a [HandleConnection], and a [HandleHolder] construct an Entity
+     * [Handle] of the right type.
+     */
+    suspend fun createHandle(
+        handleName: String,
+        handleSpec: Plan.HandleConnection,
+        holder: HandleHolder
+    ) = when (handleSpec.type) {
+        is SingletonType<*> ->
+            entityHandleManager().createSingletonHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        is CollectionType<*> ->
+            entityHandleManager().createSetHandle(
+                holder,
+                handleName,
+                handleSpec.storageKey,
+                handleSpec.type.toSchema()
+            )
+        else -> throw Exception("Unknown type ${handleSpec.type}")
```

Comment:
Please don't ever throw base `Exception`.

This is a prime location for something like `IllegalArgumentException`.